File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    answer = []
    nickname_dict = {}
    result = []
    do_tb = ['님이 들어왔습니다.', '님이 나갔습니다.']
    while record:
        line = record.pop().split()
        if line[0] == 'Enter' or line[0] == 'Change':
            if line[1] not in nickname_dict:
                nickname_dict[line[1]] = line[2]
        
        if line[0] != 'Change':
            do = 0 if line[0] == 'Enter' else 1
            result.append([line[1], do])
    
    while result:
        uid, do = result.pop()
        answer.append(nickname_dict[uid] + do_tb[do])
    return answer

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.ArrayList;
import java.util.HashMap;

class Solution {
    public String[] solution(String[] record) {
        HashMap<String, String> nameMap = new HashMap<>();
        ArrayList<String[]> temp = new ArrayList<>();
        
        for (String r : record) {
            String[] data = r.split(" ");
            String uid = data[1];
            switch (data[0]) {
                case "Enter":
                    temp.add(new String[]{uid, "님이 들어왔습니다."});
                    nameMap.put(uid, data[2]);
                    break;
                case "Leave":
                    temp.add(new String[]{uid, "님이 나갔습니다."});
                    break;
                case "Change":
                    nameMap.put(uid, data[2]);
                    break;
            }
        }
        
        int n = temp.size();
        String[] answer = new String[n];
        for (int i=0; i<n; i++) {
            String[] t = temp.get(i);
            answer[i] = nameMap.get(t[0]) + t[1];
        }
        return answer;
    }
}